gammapy/cube/psf_kernel.py

```diff
@@ -0,0 +1,310 @@
+        the input table PSF
+    geom : `~gammapy.maps.MapGeom`
+        the target geometry. The PSF kernel will be centered on the spatial center.
+    normalize : bool
+        normalize the PSF kernel (per energy)
+    factor : int
+        the oversample factor to compute the PSF
+    """
+    # First upsample spatial geom
+    upsampled_image_geom = geom.to_image().upsample(factor)
+
+    # get center coordinate
+    center_coord = upsampled_image_geom.center_coord * u.deg
+
+    # get coordinates
+    map_c = upsampled_image_geom.get_coord()
+
+    # get solid angles
+    solid_angles = upsampled_image_geom.solid_angle()
+
+    # compute distances to map center
+    rads = angular_separation(center_coord[0], center_coord[1], map_c.lon * u.deg, map_c.lat * u.deg)
+
+    vals = table_psf.evaluate(rad=rads).reshape(solid_angles.shape) * solid_angles
+
+    # Create map
+    kernel_map = Map.from_geom(geom=upsampled_image_geom.to_cube(axes=geom.axes),
+                               unit='')
+
+    # loop over images and fill map
+    for img, idx in kernel_map.iter_by_image():
+        img += vals.value
+
+    # downsample the psf kernel map. Take the average
+    kernel_map = kernel_map.downsample(factor, preserve_counts=False)
+
+    if normalize:
+        # normalize each image in map
+        for img, idx in kernel_map.iter_by_image():
+            norm = np.sum(img)
+            img /= norm
+
+    return kernel_map
+
+
+def energy_dependent_table_psf_to_kernel_map(table_psf, geom, normalize=True, factor=4):
+    """Compute an energy dependent PSF kernel on a given MapGeom.
+
+    The PSF is estimated by oversampling defined by a given factor.
+
+    Parameters
+    ----------
+    table_psf : `~gammapy.irf.EnergyDependentTablePSF`
+        the input table PSF
+    geom : `~gammapy.maps.MapGeom`
+        the target geometry. The PSF kernel will be centered on the spatial centre.
+        the geometry axes should contain an energy MapAxis.
+    normalize : bool
+        normalize the PSF kernel (per energy)
+    factor : int
+        the oversample factor to compute the PSF
+    """
+    # Find energy axis in geom
+    energy_idx = -1
+    for i, axis in enumerate(geom.axes):
+        if axis.type is 'energy':
+            energy_idx = i
+
+    if energy_idx == -1:
+        raise ValueError("No energy axis in target geometry for PSFKernel")
+
+    energy_unit = u.Unit(geom.axes[energy_idx].unit)
+
+    # TODO: change the logic to support non-regular geometry. This would allow energy dependent sizes for the kernel.
+    if geom.is_regular is False:
+        raise ValueError("Non regular geometries non supported yet.")
+
+    # First upsample spatial geom
+    upsampled_image_geom = geom.to_image().upsample(factor)
+
+    # get center coordinate
+    center_coord = upsampled_image_geom.center_coord * u.deg
+
+    # get coordinates
+    map_c = upsampled_image_geom.get_coord()
+
+    # get solid angles
```

Comment:
Is it really useful to do this complicated PSF evaluation and solid angle weighting?
My suggestion would have been to just use a pixel scale, like we did so far in Gammapy Kernel computations.
Example: https://github.com/gammapy/gammapy/pull/1388/files#diff-cde80870ee7dc037d5c9a8863a870762R76

If you want to put this computation, please don't duplicate it twice. do it just in one energy band, and from the energy-dependent method, just loop over energy bands and call the other method.

Reply:
I think this computation is not really complex in terms of code, although likely heavier than the previous methods used for kernel computation. The computing time is far from being a bottleneck anyway.
The interesting point to me is that using the `~gammapy.maps` framework potentially offers real power with simple code. e.g. if we want to have energy dependent kernel size, the change should be minimal.